runtime/lib/capi/ExecutionContext.hpp

```diff
@@ -157,112 +177,232 @@ class SharedLibraryManager final {
+    std::shared_ptr<QuantumDevice> rtd_qdevice{nullptr};
+
+    bool tape_recording{false};
 
-    // Helper methods
-    void update_device_pl2runtime(std::string &pl_name, const std::string &rtd_name)
+    void _complete_dylib_os_extension(std::string &rtd_lib, const std::string &name) noexcept
     {
 #ifdef __linux__
-        pl_name = "librtd_" + rtd_name + ".so";
+        rtd_lib = "librtd_" + name + ".so";
 #elif defined(__APPLE__)
-        pl_name = "librtd_" + rtd_name + ".dylib";
+        rtd_lib = "librtd_" + name + ".dylib";
 #endif
     }
 
+    void _pl2runtime_device_info(std::string &rtd_lib, std::string &rtd_name) noexcept
+    {
+        // The following if-elif is required for C++ tests where these backend devices
+        // are linked in the interface library of the runtime. (check runtime/CMakeLists.txt)
+        // Besides, this provides support for runtime device (RTD) libraries added to the system
+        // path. This maintains backward compatibility for specifying a device using its name.
+        // TODO: This support may need to be removed after updating the C++ unit tests.
+        if (rtd_lib == "lightning.qubit" || rtd_lib == "lightning.kokkos") {
+            rtd_name =
+                (rtd_lib == "lightning.qubit") ? "LightningSimulator" : "LightningKokkosSimulator";
+            _complete_dylib_os_extension(rtd_lib, "lightning");
+        }
+        else if (rtd_lib == "braket.aws.qubit" || rtd_lib == "braket.local.qubit") {
+            rtd_name = "OpenQasmDevice";
+            _complete_dylib_os_extension(rtd_lib, "openqasm");
+        }
+    }
+
   public:
-    explicit ExecutionContext() { _driver_mm_ptr = std::make_unique<MemoryManager>(); };
+    explicit RTDeviceType(std::string _rtd_lib, std::string _rtd_name = {},
+                          std::string _rtd_kwargs = {})
+        : rtd_lib(std::move(_rtd_lib)), rtd_name(std::move(_rtd_name)),
+          rtd_kwargs(std::move(_rtd_kwargs))
+    {
+        _pl2runtime_device_info(rtd_lib, rtd_name);
+    }
 
-    ~ExecutionContext()
+    explicit RTDeviceType(std::string_view _rtd_lib, std::string_view _rtd_name,
+                          std::string_view _rtd_kwargs)
+        : rtd_lib(_rtd_lib), rtd_name(_rtd_name), rtd_kwargs(_rtd_kwargs)
     {
-        _device_ptr.reset(nullptr);
-        _driver_mm_ptr.reset(nullptr);
-        _device_so_ptr.reset(nullptr);
-        _py_guard.reset(nullptr);
-    };
+        _pl2runtime_device_info(rtd_lib, rtd_name);
+    }
+
+    ~RTDeviceType() = default;
+
+    auto operator==(const RTDeviceType &other) const -> bool
+    {
+        return (this->rtd_lib == other.rtd_lib && this->rtd_name == other.rtd_name) &&
+               this->rtd_kwargs == other.rtd_kwargs;
+    }
+
+    [[nodiscard]] auto getQuantumDevicePtr() -> std::shared_ptr<QuantumDevice>
+    {
+        if (rtd_qdevice.get() != nullptr) {
+            return rtd_qdevice;
+        }
+
+        std::unique_ptr<SharedLibraryManager> rtd_dylib =
```

Comment:
Sure, `rtd_lib` is the absolute path to the device shared library (`/absolute/path/to/libdevice.so`) in string. We use `rtd_lib` to create an instance of `SharedLibraryManager`. This shared library manager is responsible to validate the library path, open and close the shared library, and get us the symbol to the factory method of the device. The `RTDeviceType` stores the factory method as well as `rtd_lib`. But, we don't really need to keep the share library manager around after getting the device factory method, so this will be destructed at the end of the scope of this method.

Reply:
By manager, I mean `SharedLibraryManager` and this is destructed (after fetching the factory method) at the end of this method. I cannot see any opposition here :)

Reply:
So the problem I see is with the statement that "the `SharedLibraryManager` is responsible to open and _close_ the shared library". If the `SharedLibraryManager` is destroyed right after fetching the factory method, how can it be responsible to close the shared library (whose lifetime needs to be much longer than that of the manager)?

A different, related question would be to ask, how and when is the shared library closed right now?

Reply:
We don't close the loaded library when we build the backend devices with OPENMP on Linux due to some conflicts between the OPENMP threads and dlclose. But on macOS, I ran into issues with destructing the SLM instance at the end of the method. Thus, I changed the scope to be destructed by the `RTDevice` destructor to get the support on macOS.